add @with macro

marius311 opened the pull request:
I needed to use a with statement with a ContextManager, so I went ahead an implemented the suggestion from https://github.com/JuliaPy/PyCall.jl/issues/154#issuecomment-117811984. Let me know if this is something you'd like to have, I'm also happy to make any further changes.

stevengj commented on src/PyCall.jl:
```diff
@@ -440,6 +440,59 @@ end
+    @assert as==:as "usage: @with EXPR as VAR BLOCK."
+    _with(EXPR,VAR,BLOCK)
+end
+macro with(EXPR,BLOCK)
```
I'd prefer `@pywith` to make it clear that this is for Python objects.

stevengj commented on src/PyCall.jl:
```diff
@@ -440,6 +440,59 @@ end
+    quote
+        mgr = $(esc(EXPR))
+        exit = pybuiltin("type")(mgr)[:__exit__]  # Not calling it yet
+        value = pybuiltin("type")(mgr)[:__enter__](mgr)
```
We should probably define
```jl
pytypeof(o::PyObject) = PyObject(@pycheckn ccall(@pysym(:PyObject_Type), PyPtr, (PyPtr,), o))
```
for these kinds of queries.  And probably use `mgrT = pytypeof(mgr)` to avoid calling it twice.

Use `["__exit__"]` and `["__enter__"]` to circumvent the type-introspection process, since we know we want PyObjects here.

Probably there should be a way to specify the type of object you want for `value`, by e.g. `@pywith expr as var::T ...`.   `T` would default to `PyAny` if `VAR` is specified, or `PyObject` if `VAR==nothing` (since in that case we don't need to convert the `__enter__` result to a native Julia type).  The you would call `value = pycall(mgrT["__enter__"], T, mgr)` to get the desired value type.

marius311 replied:
Good points, had not thought of that.

stevengj commented on src/PyCall.jl:
```diff
@@ -440,6 +440,59 @@ end
+            catch err
+                # The exceptional case is handled here
+                exc = false
+                if !(pybuiltin("bool")(exit(mgr, pyimport(:sys)[:exc_info]()...)))
```
What does the `bool` call do here?

marius311 replied:
Just that Python interprets None as False, and all sorts of other cases, but Julia doesn't. This makes it work consistently in Julia.

marius311 replied:
Maybe there's already a builtin way in PyCall of doing this kind of thing I don't know about?

stevengj replied:
I was hoping you could do `@pycall exit(mgr, pyimport(:sys)[:exc_info]()...)::Bool` and it would work, but this expects an integer type.

It would be better if the `Bool` conversion were changed to use `PyObject_IsTrue`, and then you could use `::Bool`.

stevengj replied:
I've updated the latest PyCall master to use `PyObject_IsTrue`, so you can just do `!pycall(exit, Bool, mgr, pyimport(:sys)[:exc_info]()...)`

stevengj commented:
Seems like a reasonable feature to have, since it is hard for users to do the enter/exit things themselves.

marius311 commented:
Cool, thanks. I'll try to tackle some of these changes over the weekend.

marius311 commented:
Ok, I added all the changes you mentioned. Also, I realized if the context manager returns a convertable type, this won't work either, so I made it so you can put `::PyObject` after the `EXPR` part too. This was in fact the thing I originally wanted to do, wihch now looks like, 

```julia
@pywith seaborn.plotting_context("paper")::PyObject begin
    plot([1,2,3])
end
```

the `::PyObject` part needed since `seaborn.plotting_context("paper")` returns a Dict (which is also a context manager, but would otherwise be converted to a Julia Dict before our macro gets to call `__enter__`, etc..)


Actually, as this kind  of thing seems to come up often and I thought the syntax was convenient, I split that off into its own macro `@pycall` which turns `@pycall func(args...)::T` into `pycall(func,T,args...)` which exists on its own (and which is used by this `@pywith` macro) . That's included in this PR, but I can take it out if there's problems and just leave the functionality inside `@pywith`.

stevengj commented on src/PyCall.jl:
```diff
@@ -576,6 +636,25 @@ pycall(o::Union{PyObject,PyPtr}, ::Type{PyAny}, args...; kwargs...) =
+Convenience macro which turns `func(args...)::T` into pycall(func, T, args...)
+"""
+macro pycall(ex)
+    @assert (isa(ex,Expr) && (ex.head==:(::)) 
```
This should just be a standard exception (probably `ArgumentError`), not an assertion failure.  Assertion failures should only be used when failure would indicate a bug in PyCall itself.

stevengj commented on src/PyCall.jl:
```diff
@@ -576,6 +636,25 @@ pycall(o::Union{PyObject,PyPtr}, ::Type{PyAny}, args...; kwargs...) =
+            && isa(ex.args[1],Expr) && (ex.args[1].head==:call)) "Usage: @pycall f(args...)::T"
+    func = ex.args[1].args[1]
+    args, kwargs = ex.args[1].args[2:end], []
+    if isa(args[1],Expr) && (args[1].head==:parameters)
```
You can use `Meta.isexpr(args[1], :parameters)` for tests like this.

stevengj commented:
`@pycall` sounds reasonable to me.

Needs tests as well.

marius311 commented:
Is it bad form to do disk IO from a test? `open` is about the only context manager in the Python standard library I can think of.

stevengj commented:
You can test with a `tempname()` file, and delete it after you are done.

stevengj commented:
Bump.  Would be great to merge this given tests.

marius311 commented:
Yes, sorry, been on my todo list to finish this, I'll try to take a look soon!

marius311 commented on src/PyCall.jl:
```diff
@@ -438,6 +438,73 @@ end
+                exit(mgr, nothing, nothing, nothing)
+            end
+        end
+        nothing
```
The reasoning here btw was that Python with-statements don't return a value, so this is for consistency. You could certainly let them return something in Julia, but then you have to make a decision what to return in case there was an error that was swallowed up by the context manager's exit function. I figured this was just simpler.

stevengj commented on test/runtests.jl:
```diff
@@ -383,3 +383,33 @@ end
+        open(readstring,fname)=="test"
+    end
+finally
+    rm(fname)
```
`rm(fname, force=true)` in case there was an error creating the file.

marius311 commented:
Just remembered this, I'm not sure if you're waiting on more from me here, but I think its ready (the test failure above was just random, I think)

stevengj commented:
Looks good, just needs a rebase to be merged.

marius311 commented:
Sounds good. Although I don't know enough how a rebase plays with Github. Would I then open a new PR? Or do you mean a merge? (or something else?)

stevengj commented:
You do a forced push to the same branch, which will update this PR.

Ismael-VC commented:
@marius311 nowadays you should be able to fix conflicts using GitHub's web GUI:

* https://help.github.com/articles/resolving-a-merge-conflict-on-github

After that @stevengj should be able to also squash the commits prior to merging using the GitHub web GUI: 

* https://github.com/blog/2141-squash-your-commits

I hope this helps!

marius311 commented:
Thanks, I didn't know about the Github conflict fixing feature, and also for the reminder that I forgot to ping @stevengj that this was OK to merge. Anyway, it looks like there's still no conflicts despite the time that's passed since I rebased this, so I think you go ahead whenever now (I think the one build error is unrelated).

stevengj commented:
Looks good; Travis error seems unrelated.
